dex/src/main/java/io/crate/data/CompositeBatchIterator.java

```diff
@@ -0,0 +1,137 @@
+ *
+ * It will prefer loaded BatchIterator's over unloaded iterators.
+ * It will consume each iterator to it's end before consuming the next iterator in order to make sure that repeated
+ * consumption via {@link #moveToStart()} always returns the same result.
+ */
+public class CompositeBatchIterator implements BatchIterator {
+
+    private final BatchIterator[] iterators;
+    private final CompositeColumns columns;
+    private int idx = 0;
+
+    private boolean loading = false;
+
+    public CompositeBatchIterator(BatchIterator ... iterators) {
+        assert iterators.length > 0 : "Must have at least 1 iterator";
+
+        // prefer loaded iterators over unloaded to improve performance in case only a subset of data is consumed
+        Arrays.sort(iterators, Comparator.comparing(BatchIterator::allLoaded).reversed());
+        this.iterators = iterators;
+        this.columns = new CompositeColumns();
+    }
+
+    @Override
+    public Columns rowData() {
+        return columns;
+    }
+
+    @Override
+    public void moveToStart() {
+        raiseIfLoading();
+        for (BatchIterator iterator : iterators) {
+            iterator.moveToStart();
+        }
+        idx = 0;
+    }
+
+    @Override
+    public boolean moveNext() {
+        raiseIfLoading();
+        while (idx < iterators.length) {
+            BatchIterator iterator = iterators[idx];
+            if (iterator.moveNext()) {
+                return true;
+            }
+            if (iterator.allLoaded() == false) {
+                return false;
+            }
+            idx++;
+        }
+        idx = 0;
+        return false;
+    }
+
+    @Override
+    public void close() {
+        for (BatchIterator iterator : iterators) {
+            iterator.close();
+        }
+    }
+
+    @Override
+    public CompletionStage<?> loadNextBatch() {
+        if (loading) {
+            return CompletableFutures.failedFuture(new IllegalStateException("BatchIterator already loading"));
+        }
+        for (BatchIterator iterator : iterators) {
+            if (iterator.allLoaded()) {
+                continue;
+            }
+            loading = true;
+            return iterator.loadNextBatch().whenComplete((r, t) -> loading = false);
+        }
+        return CompletableFutures.failedFuture(new IllegalStateException("BatchIterator already fully loaded"));
+    }
+
+    @Override
+    public boolean allLoaded() {
+        raiseIfLoading();
+        for (BatchIterator iterator : iterators) {
+            if (iterator.allLoaded() == false) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void raiseIfLoading() {
+        if (loading) {
+            throw new IllegalStateException("BatchIterator already loading");
+        }
+    }
+
+    private class CompositeColumns implements Columns {
+
+        @Override
+        public Input<?> get(int index) {
+            return iterators[idx].rowData().get(index);
```

Comment:
@dobe I think this here also won't work if the inputs are retrieved upfront. Any suggestions how this could be solved?